### PR TITLE
Implement first version of non-interactive preview for desktop

### PR DESCRIPTION
**Pull request by AlexeyTsvetkov**

### CODE REVIEW

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/internal/projectUtils.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+internal fun Project.intProperty(name: String): Int {
+    val value = project.property(name)
+    return (value as? String)?.toIntOrNull()
+        ?: error("Property '$name' is not an integer: $value")
+}
+
+internal fun Project.stringProperty(name: String): String {
+    val value = project.property(name)
+    return value as? String
+        ?: error("Property '$name' is not a string: $value")
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/internal/projectUtils.kt:19]: Can it actually be non-String or null?

File: gradle-plugins/preview-rpc/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+plugins {
```
Comment (olonho) [inline gradle-plugins/preview-rpc/build.gradle.kts:1]: Why we need separate preview-rpc Gralde plugin?

File: gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/PreviewManager.kt [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    private val frameRequest = AtomicReference<FrameRequest>(null)
+    private val shouldRequestFrame = AtomicBoolean(false)
+    private val runningPreview = AtomicReference<RunningPreview>(null)
+    private val threads = arrayListOf<Thread>()
+
+    private val runPreviewThread = repeatWhileAliveThread {
+        fun startPreviewProcess(config: PreviewHostConfig): Process =
+            ProcessBuilder(
+                config.javaExecutable,
+                "-Djava.awt.headless=true",
```
Comment (olonho) [inline gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/PreviewManager.kt:64]: Nice!

File: gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/PreviewManager.kt [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    private val sendPreviewRequestThread = repeatWhileAliveThread {
+        withLivePreviewConnection {
+            val classpath = previewClasspath.get()
+            val fqName = previewFqName.get()
+            val frameSize = previewFrameSize.get()
+
+            if (classpath != null && frameSize != null && fqName != null) {
+                val request = FrameRequest(fqName, frameSize.width, frameSize.height)
+                if (shouldRequestFrame.get() && frameRequest.get() == null) {
+                    if (shouldRequestFrame.compareAndSet(true, false)) {
```
Comment (olonho) [inline gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/PreviewManager.kt:95]: Maybe atomics are overkill here :), but we could go with them.

File: gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/PreviewManager.kt [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        runningPreview.get()?.let { (connection, process) ->
+            connection.close()
+            if (!process.waitFor(5, TimeUnit.SECONDS)) {
+                // todo: check exit code
+                process.destroyForcibly()
+            }
+
+        }
+        log { "CLOSING PREVIEW SOCKET" }
+        previewSocket.close()
```
Comment (olonho) [inline gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/PreviewManager.kt:163]: Shouldn't those be in `try` block?

File: gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/commands.kt [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                receiveUtf8StringData { onPreviewFqName(it) }
+            Command.Type.PREVIEW_CONFIG -> {
+                val javaExecutable = args[0]
+                receiveUtf8StringData { hostClasspath ->
+                    val config = PreviewHostConfig(javaExecutable = javaExecutable, hostClasspath = hostClasspath)
+                    onPreviewHostConfig(config)
+                }
+            }
+            else -> {
+                // todo
```
Comment (olonho) [inline gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/commands.kt:65]: Maybe throw?

File: gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/utils.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        } catch (e: IOException) {
+            Thread.sleep(trialDelay)
+        }
+    }
+
+    return null
+}
+
+internal val localhost: InetAddress
+    get() = InetAddress.getByName("127.0.0.1")
```
Comment (olonho) [inline gradle-plugins/preview-rpc/src/main/kotlin/org/jetbrains/compose/desktop/ui/tooling/preview/rpc/utils.kt:30]: `InetAddress.getLoopbackAddress()` ?

File: idea-plugin/build.gradle.kts [hunk 8]
```diff
@@ -58,4 +62,9 @@ tasks {
         ideVersions(properties("plugin.verifier.ide.versions"))
         downloadDirectory("${project.buildDir}/pluginVerifier/ides")
     }
+
+    runIde {
```
Comment (olonho) [inline idea-plugin/build.gradle.kts:66]: Why this change?

File: idea-plugin/examples/desktop-project/build.gradle.kts [hunk 9]
```diff
@@ -13,6 +13,7 @@ repositories {
 
 dependencies {
     implementation("org.jetbrains.kotlinx:kotlinx-collections-immutable:0.3.4")
+    implementation(compose.uiTooling)
```
Comment (olonho) [inline idea-plugin/examples/desktop-project/build.gradle.kts:16]: Hmm, could we avoid this dependency, please?

File: gradle-plugins/preview-rpc/build.gradle.kts (hunk 2 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/preview-rpc/build.gradle.kts:1]: It's not a plugin, it's a separate module, that is used by both the Gradle plugin and the Intellij plugin. It needs to be separate module, because I don't want to introduce a compile/runtime dependency between the Gradle plugin and the Intellij plugin.

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/internal/projectUtils.kt (hunk 1 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/internal/projectUtils.kt]: I removed this code

File: idea-plugin/build.gradle.kts (hunk 8 above)
Comment (AlexeyTsvetkov) [reply idea-plugin/build.gradle.kts:66]: It was convenient to always publish the latest Gradle plugin for dev iterations, but now thinking about it, I'm not sure if it's such a universally good idea, so I will revert the change

**Review by olonho — APPROVED** [review]
